beps/0006-scaffolder-action-rollback/README.md

```diff
@@ -31,36 +37,281 @@ Introducing the rollback to scaffolder actions provides the mean to come back to
 
 ## Motivation
 
-The aim is to help to mitigate the issue of manual clean up the partially created resources.
+Mitigate the issue of manual clean up from partially created resources during template execution failures. When a template execution fails after creating a number of resources but before completing, the system is left in an inconsistent state. Currently, there is no standardized way to clean up these partially created resources, leading to orphaned resources in external systems requiring manual cleanup.
 
 ### Goals
 
-- We will extend action's API to be able to run rollback for failed tasks
-- Rollback will be optional
+- Provide a standardized, optional rollback mechanism for scaffolder actions
+- Enable proper cleanup of resources created in third-party systems when template execution fails
+- Improve the developer experience by reducing manual intervention for cleanup after failures
+- Provide a clear indicator of which actions have rollbacks in the UI (/create/actions)
+- Provide override options when possible
+- Ensure auditing is a first class concern
 
 ### Non-Goals
 
-We will not cover all built-in actions with a rollback functionality
+- Modifying existing actions to implement rollback (this proposal only provides the framework)
+- Adding automatic rollback triggers for all possible error scenarios
 
 ## Proposal
 
 Rollback is going to be performed:
 
-- when user manually decide to perform this action.
-- task has to be recovered and TaskRecoverStrategy set to 'rollback'
+- on template execution failure
+- for each step when rollback is enabled and the correlating action includes a rollback function
+- and possibly incorporate when `TaskRecoverStrategy` set to 'rollback' (needs more research)
 
 ## Design Details
 
-We are going to introduce an extra function in action API which will look like:
+### Auditing and Permissions
+
+1. Implement the AuditorService for key operations
+1. Add a rollback permission definition with permission checks in the NunjucksWorkflowRunner
+1. Add a specific rule for rollback operations to `scaffolderActionRules` ([ref](https://github.com/backstage/backstage/blob/85df833fe33c45cf2ad8e0322dafb118b89a5ea8/plugins/scaffolder-backend/src/service/rules.ts#L133-L138))
+
+### Securing Rollbacks
+
+We will provide a helper function that encapsulates the logic of verifying the resource and executing the rollback. Core security checks can be performed in this function to ensure the resource is
+valid and can be rolled back. There is a possible opportunity to allow others to extend this function to provide custom security checks.
+
+By default, we can handle the following:
+
+1. Verify we reached the point of setting output
+1. Parse resource details from the output and validate them against the input
+1. Execute the actual rollback with the validated resource
+
+```typescript
+export function createSecureRollback
```

Comment:
I think this method can be simplified in terms of input.
Tasks should be idempotent, and for that we use checkpoints.
I'm slowly adding them into built-in actions, you can see them in any of gitlab actions and some of github's too.

The results of checkpoints are already stored in the database.
If we need something extra for rollback, i'd store it there.

and even maybe we can do rollback in the scope of checkpoint?

Some of the actions are huge, like for example 'github:environment:create'.

There we create multiples things: users, policy tags, policy branches, secrets, etc.

If we will keep rollback to the checkpoint, it would be cleaner to see what to rollback.

Like an atomic stuff to create/delete.

Reply:
@benjdlambert  wdyt?

Reply:
I spent a small amount of time looking into checkpoints. I'm happy[ I was on the right track](https://github.com/backstage/backstage/pull/29196#issuecomment-2725844405). I will definitely spend time diving in on this. Looking forward to the feedback from others.